<a href="https://colab.research.google.com/github/richardjer7/Common/blob/master/Actividad_M4S2_Grupo_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Actividad M4S2

Ajuste de hiperparámetros para un modelo de clasificación médica que predice la diabetes utilizando el conjunto de datos "https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv". La tarea incluye la carga y exploración de datos, el preprocesamiento (escalado y división), el entrenamiento RandomForestClassifier, la aplicación de Grid Search y Random Search para la optimización de hiperparámetros (n_estimators, max_depth, min_samples_split), la evaluación y comparación del rendimiento (accuracy, F1-score, computational efficiency) de los modelos obtenidos con ambos métodos de búsqueda y la elaboración de una reflexión final sobre el proceso.

## Carga y exploración de datos

### Subtask:
Cargar el conjunto de datos "Pima Indians Diabetes Dataset" desde la URL proporcionada, explorar su estructura, variables y describir brevemente su relación con la predicción de diabetes.


**Reasoning**:
Se importará la librería pandas y se cargará el conjunto de datos desde la URL proporcionada en un DataFrame, asignando nombres a las columnas. Luego, se mostrarán las primeras filas, información general y un resumen estadístico del DataFrame para explorar los datos. Finalmente, se contará el número de valores en la columna 'resultado'.



In [ ]:
import pandas as pd

# Nombres de las columnas
nombres_columnas = ['embarazos', 'glucosa', 'presion_arterial', 'espesor_piel', 'insulina', 'bmi', 'pedigree', 'edad', 'resultado']

# Cargar el conjunto de datos
df = pd.read_csv("https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv", names=nombres_columnas)

# Mostrar las primeras 5 filas
print("Primeras 5 filas del DataFrame:")
display(df.head())

# Mostrar información del DataFrame
print("\nInformación del DataFrame:")
display(df.info())

# Mostrar resumen estadístico
print("\nResumen estadístico del DataFrame:")
display(df.describe())

# Contar valores en la columna 'resultado'
print("\nConteo de valores en la columna 'resultado':")
display(df['resultado'].value_counts())

Primeras 5 filas del DataFrame:


embarazos  glucosa  presion_arterial  espesor_piel  insulina   bmi  \
0          6      148                72            35         0  33.6   
1          1       85                66            29         0  26.6   
2          8      183                64             0         0  23.3   
3          1       89                66            23        94  28.1   
4          0      137                40            35       168  43.1   

   pedigree  edad  resultado  
0     0.627    50          1  
1     0.351    31          0  
2     0.672    32          1  
3     0.167    21          0  
4     2.288    33          1


Información del DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   embarazos         768 non-null    int64  
 1   glucosa           768 non-null    int64  
 2   presion_arterial  768 non-null    int64  
 3   espesor_piel      768 non-null    int64  
 4   insulina          768 non-null    int64  
 5   bmi               768 non-null    float64
 6   pedigree          768 non-null    float64
 7   edad              768 non-null    int64  
 8   resultado         768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


None


Resumen estadístico del DataFrame:


embarazos     glucosa  presion_arterial  espesor_piel    insulina  \
count  768.000000  768.000000        768.000000    768.000000  768.000000   
mean     3.845052  120.894531         69.105469     20.536458   79.799479   
std      3.369578   31.972618         19.355807     15.952218  115.244002   
min      0.000000    0.000000          0.000000      0.000000    0.000000   
25%      1.000000   99.000000         62.000000      0.000000    0.000000   
50%      3.000000  117.000000         72.000000     23.000000   30.500000   
75%      6.000000  140.250000         80.000000     32.000000  127.250000   
max     17.000000  199.000000        122.000000     99.000000  846.000000   

              bmi    pedigree        edad   resultado  
count  768.000000  768.000000  768.000000  768.000000  
mean    31.992578    0.471876   33.240885    0.348958  
std      7.884160    0.331329   11.760232    0.476951  
min      0.000000    0.078000   21.000000    0.000000  
25%     27.300000    0.243750   24.000000    0.000000  
50%     32.000000    0.372500   29.000000    0.000000  
75%     36.600000    0.626250   41.000000    1.000000  
max     67.100000    2.420000   81.000000    1.000000


Conteo de valores en la columna 'resultado':


resultado
0    500
1    268
Name: count, dtype: int64

## Preprocesamiento de datos

### Subtask:
Escalar las variables numéricas utilizando `StandardScaler` y dividir el conjunto de datos en conjuntos de entrenamiento y prueba.


**Reasoning**:
Separar las características y la variable objetivo, escalar las características numéricas y dividir los datos en conjuntos de entrenamiento y prueba.



In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

# Separar características (X) y variable objetivo (y)
X = df.drop('resultado', axis=1)
y = df['resultado']

# Instanciar StandardScaler
scaler = StandardScaler()

# Aplicar StandardScaler a las características
X_scaled = scaler.fit_transform(X)

# Dividir los datos escalados en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=42)

print("Dimensiones de los conjuntos de datos:")
print("X_train:", X_train.shape)
print("X_test:", X_test.shape)
print("y_train:", y_train.shape)
print("y_test:", y_test.shape)

Dimensiones de los conjuntos de datos:
X_train: (576, 8)
X_test: (192, 8)
y_train: (576,)
y_test: (192,)


## Modelo base

### Subtask:
Entrenar un `RandomForestClassifier` con los hiperparámetros por defecto y evaluar su rendimiento utilizando métricas como precisión, recall, F1-Score y AUC.


* Importar las bibliotecas necesarias para el entrenamiento y evaluación del modelo.


In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, f1_score, recall_score, roc_auc_score

* Crear una instancia y entrenamiento de un RandomForestClassifier con hiperparámetros predeterminados y realización de predicciones en el conjunto de prueba.



In [ ]:
# Instanciar un RandomForestClassifier con hiperparámetros por defecto
rf_model = RandomForestClassifier(random_state=42)

# Entrenar el modelo utilizando los datos de entrenamiento escalados
rf_model.fit(X_train, y_train)

# Realizar predicciones sobre el conjunto de prueba escalado
y_pred = rf_model.predict(X_test)

**Reasoning**:
Calculate and print the evaluation metrics for the trained Random Forest model using the test set predictions and the true test labels.



In [ ]:
# Calcular métricas de evaluación
accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
auc = roc_auc_score(y_test, y_pred)

# Imprimir las métricas
print(f"Precisión del modelo: {accuracy:.4f}")
print(f"F1-Score del modelo: {f1:.4f}")
print(f"Recall del modelo: {recall:.4f}")
print(f"AUC del modelo: {auc:.4f}")

Precisión del modelo: 0.7344
F1-Score del modelo: 0.6383
Recall del modelo: 0.6522
AUC del modelo: 0.7163


## Aplicar grid search

### Subtask:
Definir un rango de hiperparámetros para `RandomForestClassifier` y utilizar `GridSearchCV` para encontrar la mejor combinación de hiperparámetros basándose en el rendimiento del modelo.


**Reasoning**:
Import the GridSearchCV class to perform hyperparameter tuning.



In [ ]:
from sklearn.model_selection import GridSearchCV

**Reasoning**:
Define the hyperparameter grid, instantiate the RandomForestClassifier, and instantiate and fit GridSearchCV to the training data.



In [ ]:
# Definir el rango de hiperparámetros a tunear
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10]
}

# Instanciar un RandomForestClassifier
rf = RandomForestClassifier(random_state=42)

# Instanciar GridSearchCV
grid_search = GridSearchCV(estimator=rf, param_grid=param_grid, cv=5, scoring='f1', n_jobs=-1)

# Ajustar GridSearchCV a los datos de entrenamiento escalados
grid_search.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=RandomForestClassifier(random_state=42), n_jobs=-1,
             param_grid={'max_depth': [None, 10, 20, 30],
                         'min_samples_split': [2, 5, 10],
                         'n_estimators': [100, 200, 300]},
             scoring='f1')

**Reasoning**:
Print the best hyperparameters and the best F1-score found by GridSearchCV, and store the best estimator.



In [ ]:
# Imprimir los mejores hiperparámetros
print("Mejores hiperparámetros encontrados por GridSearchCV:")
print(grid_search.best_params_)

# Imprimir el mejor score (F1-score) obtenido por GridSearchCV
print("\nMejor F1-score obtenido por GridSearchCV:")
print(grid_search.best_score_)

# Almacenar el mejor modelo
best_grid_model = grid_search.best_estimator_

Mejores hiperparámetros encontrados por GridSearchCV:
{'max_depth': 10, 'min_samples_split': 5, 'n_estimators': 100}

Mejor F1-score obtenido por GridSearchCV:
0.6431876359275076


## Aplicar random search

### Subtask:
Definir un rango más amplio de hiperparámetros y utilizar `RandomizedSearchCV` con un número específico de iteraciones para encontrar una buena combinación de hiperparámetros.


**Reasoning**:
Import the necessary class for randomized search and define a broader hyperparameter distribution for the RandomForestClassifier.



In [ ]:
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import randint

# Definir un rango más amplio de hiperparámetros
param_dist = {
    'n_estimators': randint(low=50, high=500),
    'max_depth': [None] + list(randint(low=5, high=50).rvs(10)), # None and 10 random integers between 5 and 50
    'min_samples_split': randint(low=2, high=20)
}

**Reasoning**:
Instantiate a RandomForestClassifier and RandomizedSearchCV, then fit the search object to the training data.



In [ ]:
# Instanciar un RandomForestClassifier con un random_state para reproducibilidad
rf = RandomForestClassifier(random_state=42)

# Instanciar RandomizedSearchCV
random_search = RandomizedSearchCV(estimator=rf, param_distributions=param_dist, n_iter=100, cv=5, scoring='f1', n_jobs=-1, random_state=42)

# Ajustar RandomizedSearchCV a los datos de entrenamiento escalados
random_search.fit(X_train, y_train)

RandomizedSearchCV(cv=5, estimator=RandomForestClassifier(random_state=42),
                   n_iter=100, n_jobs=-1,
                   param_distributions={'max_depth': [None, np.int64(15),
                                                      np.int64(7), np.int64(12),
                                                      np.int64(42),
                                                      np.int64(12), np.int64(9),
                                                      np.int64(41),
                                                      np.int64(22),
                                                      np.int64(21),
                                                      np.int64(5)],
                                        'min_samples_split': <scipy.stats._distn_infrastructure.rv_discrete_frozen object at 0x7ddd8328cc90>,
                                        'n_estimators': <scipy.stats._distn_infrastructure.rv_discrete_frozen object at 0x7ddd865d6910>},
                   random_state=42, scoring='f1')

**Reasoning**:
Print the best hyperparameters and best score found by RandomizedSearchCV and store the best model.



In [ ]:
# Imprimir los mejores hiperparámetros encontrados por RandomizedSearchCV
print("Mejores hiperparámetros encontrados por RandomizedSearchCV:")
print(random_search.best_params_)

# Imprimir el mejor score (F1-score) obtenido por RandomizedSearchCV
print("\nMejor F1-score obtenido por RandomizedSearchCV:")
print(random_search.best_score_)

# Almacenar el mejor modelo encontrado por RandomizedSearchCV
best_random_model = random_search.best_estimator_

Mejores hiperparámetros encontrados por RandomizedSearchCV:
{'max_depth': np.int64(12), 'min_samples_split': 8, 'n_estimators': 357}

Mejor F1-score obtenido por RandomizedSearchCV:
0.6465232646834478


## Comparación y análisis

### Subtask:
Evaluar los modelos obtenidos con Grid Search y Random Search en el conjunto de prueba, comparar sus métricas de rendimiento y los tiempos de entrenamiento.


**Reasoning**:
Import the necessary libraries for evaluation and time measurement.



In [ ]:
from sklearn.metrics import accuracy_score, f1_score, recall_score, roc_auc_score
import time

**Reasoning**:
Define a function to train and evaluate a model, then call it for each model.



In [ ]:
def train_and_evaluate_model(model, X_train, y_train, X_test, y_test, model_name):
    """Entrena y evalúa un modelo de clasificación, imprimiendo las métricas."""
    print(f"Evaluando modelo: {model_name}")

    start_time = time.time()
    model.fit(X_train, y_train)
    end_time = time.time()
    training_time = end_time - start_time
    print(f"Tiempo de entrenamiento: {training_time:.4f} segundos")

    y_pred = model.predict(X_test)

    accuracy = accuracy_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    auc = roc_auc_score(y_test, y_pred)

    print(f"Precisión: {accuracy:.4f}")
    print(f"F1-Score: {f1:.4f}")
    print(f"Recall: {recall:.4f}")
    print(f"AUC: {auc:.4f}")
    print("-" * 30)

    return {'accuracy': accuracy, 'f1': f1, 'recall': recall, 'auc': auc, 'training_time': training_time}

# Evaluar el modelo base
base_model_results = train_and_evaluate_model(rf_model, X_train, y_train, X_test, y_test, "Modelo Base (RandomForest por defecto)")

# Evaluar el mejor modelo de Grid Search
grid_search_model_results = train_and_evaluate_model(best_grid_model, X_train, y_train, X_test, y_test, "Mejor Modelo (Grid Search)")

# Evaluar el mejor modelo de Random Search
random_search_model_results = train_and_evaluate_model(best_random_model, X_train, y_train, X_test, y_test, "Mejor Modelo (Random Search)")

Evaluando modelo: Modelo Base (RandomForest por defecto)
Tiempo de entrenamiento: 0.4291 segundos
Precisión: 0.7344
F1-Score: 0.6383
Recall: 0.6522
AUC: 0.7163
------------------------------
Evaluando modelo: Mejor Modelo (Grid Search)
Tiempo de entrenamiento: 0.3089 segundos
Precisión: 0.7448
F1-Score: 0.6475
Recall: 0.6522
AUC: 0.7245
------------------------------
Evaluando modelo: Mejor Modelo (Random Search)
Tiempo de entrenamiento: 1.1570 segundos
Precisión: 0.7448
F1-Score: 0.6525
Recall: 0.6667
AUC: 0.7276
------------------------------


## Reflexión final

### Subtask:
Analizar cuál técnica de búsqueda de hiperparámetros fue más eficiente, cuál encontró el mejor modelo y reflexionar sobre posibles enfoques alternativos.


**Reasoning**:
Analizaré los resultados comparando las métricas de rendimiento y los tiempos de entrenamiento de los tres modelos para determinar la eficiencia de las técnicas de búsqueda de hiperparámetros y discutiré enfoques alternativos, como se indica en las instrucciones.



In [ ]:
import numpy as np

print("--- Análisis de los Resultados ---")

# 1. Comparar los resultados de los modelos
print("\nComparación de Métricas de Rendimiento y Tiempo de Entrenamiento:")
print("-----------------------------------------------------------------")
print(f"{'Modelo':<30} | {'Accuracy':<10} | {'F1-Score':<10} | {'Recall':<10} | {'AUC':<10} | {'Tiempo (s)':<15}")
print("-" * 100)
print(f"{'Modelo Base (RandomForest)':<30} | {base_model_results['accuracy']:<10.4f} | {base_model_results['f1']:<10.4f} | {base_model_results['recall']:<10.4f} | {base_model_results['auc']:<10.4f} | {base_model_results['training_time']:<15.4f}")
print(f"{'Mejor Modelo (Grid Search)':<30} | {grid_search_model_results['accuracy']:<10.4f} | {grid_search_model_results['f1']:<10.4f} | {grid_search_model_results['recall']:<10.4f} | {grid_search_model_results['auc']:<10.4f} | {grid_search_model_results['training_time']:<15.4f}")
print(f"{'Mejor Modelo (Random Search)':<30} | {random_search_model_results['accuracy']:<10.4f} | {random_search_model_results['f1']:<10.4f} | {random_search_model_results['recall']:<10.4f} | {random_search_model_results['auc']:<10.4f} | {random_search_model_results['training_time']:<15.4f}")

# 2. Analizar qué técnica encontró el mejor modelo
print("\nAnálisis del Mejor Modelo:")
print("---------------------------")

best_f1_grid = grid_search_model_results['f1']
best_f1_random = random_search_model_results['f1']

if best_f1_random > best_f1_grid:
    print("Random Search encontró un modelo con un F1-Score ligeramente mejor en el conjunto de prueba.")
elif best_f1_grid > best_f1_random:
     print("Grid Search encontró un modelo con un F1-Score ligeramente mejor en el conjunto de prueba.")
else:
    print("Grid Search y Random Search encontraron modelos con el mismo F1-Score en el conjunto de prueba.")

# También considerar otras métricas
if random_search_model_results['auc'] > grid_search_model_results['auc']:
    print("Random Search también obtuvo un AUC ligeramente mayor.")
elif grid_search_model_results['auc'] > random_search_model_results['auc']:
    print("Grid Search también obtuvo un AUC ligeramente mayor.")


# 3. Discutir la eficiencia computacional
print("\nEficiencia Computacional:")
print("-------------------------")
print(f"Tiempo de entrenamiento del modelo base: {base_model_results['training_time']:.4f} segundos")
print(f"Tiempo de entrenamiento del mejor modelo de Grid Search: {grid_search_model_results['training_time']:.4f} segundos")
print(f"Tiempo de entrenamiento del mejor modelo de Random Search: {random_search_model_results['training_time']:.4f} segundos")

if grid_search_model_results['training_time'] < random_search_model_results['training_time']:
    print("Grid Search fue considerablemente más rápido que Random Search en este caso particular.")
elif random_search_model_results['training_time'] < grid_search_model_results['training_time']:
     print("Random Search fue más rápido que Grid Search en este caso particular.")
else:
    print("Los tiempos de entrenamiento fueron similares.")

# 4. Reflexión sobre por qué una técnica superó a la otra (o no)
print("\nReflexión:")
print("-----------")
print("En este caso, Random Search logró un F1-Score y AUC marginalmente superiores en el conjunto de prueba, lo que sugiere que exploró combinaciones de hiperparámetros que llevaron a un modelo ligeramente mejor en este espacio de búsqueda y con este número de iteraciones.")
print("A pesar de encontrar un modelo marginalmente mejor, Random Search tuvo un tiempo de entrenamiento significativamente mayor que Grid Search.")
print("La superioridad de una técnica sobre otra puede depender del tamaño y la naturaleza del espacio de búsqueda, el número de iteraciones (en el caso de Random Search) y la distribución de los hiperparámetros óptimos dentro de ese espacio.")
print("Grid Search explora exhaustivamente todas las combinaciones en una cuadrícula definida, mientras que Random Search muestrea aleatoriamente, lo que puede ser más eficiente en espacios de alta dimensión o cuando los hiperparámetros óptimos no están alineados con una cuadrícula predefinida.")

# 5. Sugerir posibles enfoques alternativos
print("\nEnfoques Alternativos para Optimización de Hiperparámetros:")
print("----------------------------------------------------------")
print("- **Optimización Bayesiana:** Utiliza resultados de evaluaciones previas para guiar la búsqueda hacia regiones prometedoras del espacio de hiperparámetros.")
print("- **Optimización basada en Gradientes:** (Aunque menos común para hiperparámetros de modelos de árboles) Explora el espacio de hiperparámetros utilizando información de gradientes.")
print("- **Algoritmos Genéticos:** Emplean principios de selección natural para encontrar combinaciones óptimas de hiperparámetros.")
print("- **Bibliotecas de Optimización Automática (AutoML):** Herramientas más completas que automatizan el proceso de selección de modelo e hiperparámetros (ej. Auto-Sklearn, TPOT).")

# 6. Resumen conciso
print("\n--- Resumen ---")
print("En este análisis, comparamos los modelos de RandomForest obtenidos con hiperparámetros por defecto, Grid Search y Random Search.")
print("Grid Search fue significativamente más rápido en el entrenamiento, mientras que Random Search encontró un modelo con métricas de rendimiento (F1-Score, AUC) marginalmente mejores en el conjunto de prueba.")
print("Esto sugiere que, aunque Random Search puede encontrar modelos ligeramente mejores al explorar un espacio de búsqueda más amplio, puede ser menos eficiente computacionalmente dependiendo del número de iteraciones y el espacio definido.")
print("Existen enfoques alternativos como la Optimización Bayesiana y las herramientas de AutoML que podrían explorarse para una optimización más eficiente y potencialmente mejor.")

--- Análisis de los Resultados ---

Comparación de Métricas de Rendimiento y Tiempo de Entrenamiento:
-----------------------------------------------------------------
Modelo                         | Accuracy   | F1-Score   | Recall     | AUC        | Tiempo (s)     
----------------------------------------------------------------------------------------------------
Modelo Base (RandomForest)     | 0.7344     | 0.6383     | 0.6522     | 0.7163     | 0.4291         
Mejor Modelo (Grid Search)     | 0.7448     | 0.6475     | 0.6522     | 0.7245     | 0.3089         
Mejor Modelo (Random Search)   | 0.7448     | 0.6525     | 0.6667     | 0.7276     | 1.1570         

Análisis del Mejor Modelo:
---------------------------
Random Search encontró un modelo con un F1-Score ligeramente mejor en el conjunto de prueba.
Random Search también obtuvo un AUC ligeramente mayor.

Eficiencia Computacional:
-------------------------
Tiempo de entrenamiento del modelo base: 0.4291 segundos
Tiempo de ent

## Resumen:

### Hallazgos Clave del Análisis de Datos

* El conjunto de datos contiene 768 instancias con 8 características y una variable objetivo binaria ('resultado') que indica la presencia (1) o ausencia (0) de diabetes. Existe un desbalance de clases con 500 instancias de la clase 0 y 268 de la clase 1.
* Varias características como 'glucosa', 'presion_arterial', 'espesor_piel', 'insulina' y 'bmi' contenían valores de 0, lo que podría representar datos faltantes.
* El `RandomForestClassifier` base con hiperparámetros por defecto logró un F1-Score de aproximadamente 0.6383 y un AUC de 0.7163 en el conjunto de prueba, con un tiempo de entrenamiento de alrededor de 0.4291 segundos.
* Grid Search encontró los mejores hiperparámetros `{'max_depth': 10, 'min_samples_split': 5, 'n_estimators': 100}`, resultando en un modelo con un F1-Score de aproximadamente 0.6475 y un AUC de 0.7245 en el conjunto de prueba. El tiempo de entrenamiento para este modelo fue significativamente más rápido, alrededor de 0.3089 segundos.
* Random Search, con 100 iteraciones y una distribución de parámetros más amplia, identificó los mejores hiperparámetros `{'max_depth': 12, 'min_samples_split': 8, 'n_estimators': 357}`, obteniendo un modelo con un F1-Score (0.6525) y AUC (0.7276) ligeramente mejores en el conjunto de prueba en comparación con Grid Search. Sin embargo, su tiempo de entrenamiento fue considerablemente más largo, alrededor de 1.1570 segundos.
* Ambos métodos de ajuste de hiperparámetros mejoraron las métricas de rendimiento del modelo (Accuracy, F1-Score, AUC) en comparación con el modelo base.

### Ideas o Próximos Pasos

* Si bien Random Search encontró un modelo marginalmente mejor en términos de F1-Score y AUC, Grid Search fue significativamente más eficiente computacionalmente en este escenario y espacio de parámetros específico. La elección entre ellos depende del equilibrio deseado entre la ganancia potencial de rendimiento y el costo computacional.
* Explorar técnicas alternativas de optimización de hiperparámetros, como la Optimización Bayesiana, que pueden buscar de manera más eficiente en espacios de parámetros complejos, encontrando potencialmente mejores modelos que Grid Search o Random Search dentro de un plazo razonable.